src/vs/editor/contrib/hover/browser/contentHover.ts

```diff
@@ -437,252 +442,385 @@ class ContentHoverVisibleData {
+			return;
 		}
-	}
-
-	public onContentsChanged(): void {
+		this._setHoverData(undefined);
+		this._resizableNode.maxSize = new dom.Dimension(Infinity, Infinity);
+		this._resizableNode.clearSashHoverState();
+		this._editor.removeContentWidget(this);
+		this._hoverFocusedKey.set(false);
 		this._editor.layoutContentWidget(this);
-		this._hover.onContentsChanged();
-
-		const scrollDimensions = this._hover.scrollbar.getScrollDimensions();
-		const hasHorizontalScrollbar = (scrollDimensions.scrollWidth > scrollDimensions.width);
-		if (hasHorizontalScrollbar) {
-			// There is just a horizontal scrollbar
-			const extraBottomPadding = `${this._hover.scrollbar.options.horizontalScrollbarSize}px`;
-			if (this._hover.contentsDomNode.style.paddingBottom !== extraBottomPadding) {
-				this._hover.contentsDomNode.style.paddingBottom = extraBottomPadding;
-				this._editor.layoutContentWidget(this);
-				this._hover.onContentsChanged();
-			}
+		if (this._hoverData.stoleFocus) {
+			this._editor.focus();
 		}
 	}
 
-	public clear(): void {
-		this._hover.contentsDomNode.textContent = '';
+	private _setPersistedHoverDimensionsOrRenderNormally(): void {
+		// Added because otherwise the initial size of the hover content is smaller than should be
+		const layoutInfo = this._editor.getLayoutInfo();
+		this._resizableNode.layout(layoutInfo.height, layoutInfo.width);
+		this._setHoverWidgetDimensions('auto', 'auto');
+	}
+
+	private _setContainerAbsolutePosition(top: number, left: number): void {
+		const containerDomNode = this._hoverWidget.containerDomNode;
+		containerDomNode.style.top = top + 'px';
+		containerDomNode.style.left = left + 'px';
+	}
+
+	private _adjustHoverHeightForScrollbar(height: number) {
+		const containerDomNode = this._hoverWidget.containerDomNode;
+		const contentsDomNode = this._hoverWidget.contentsDomNode;
+		const maxRenderingHeight = this._findMaximumRenderingHeight() ?? Infinity;
+		this._setContainerDomNodeDimensions(containerDomNode.clientWidth, Math.min(maxRenderingHeight, height));
+		this._setContentsDomNodeDimensions(contentsDomNode.clientWidth, Math.min(maxRenderingHeight, height - SCROLLBAR_WIDTH));
+	}
+
+	public onContentsChanged(): void {
+		this._setPersistedHoverDimensionsOrRenderNormally();
+		const containerDomNode = this._hoverWidget.containerDomNode;
+		const clientHeight = containerDomNode.clientHeight;
+		const clientWidth = containerDomNode.clientWidth;
+		const totalBorderWidth = 2 * SASH_WIDTH_MINUS_BORDER;
+		this._resizableNode.layout(clientHeight + totalBorderWidth, clientWidth + totalBorderWidth);
+		this._setContainerAbsolutePosition(SASH_WIDTH_MINUS_BORDER - 1, SASH_WIDTH_MINUS_BORDER - 1);
+		if (this._hasHorizontalScrollbar()) {
+			this._adjustContentsBottomPadding();
+			this._adjustHoverHeightForScrollbar(clientHeight);
+		}
+		this._layoutContentWidget();
 	}
 
 	public focus(): void {
-		this._hover.containerDomNode.focus();
+		this._hoverWidget.containerDomNode.focus();
```

Comment:
suggestion: there are a bunch of renames that make this PR larger than needed (`_hoverData`, `ResizableHoverWidget`, `_hoverWidget` etc.) No feelings about the changes per se but I would suggest you do them in a separate PR so that this PR stays focused to resizing and so that the GIT history is more precise